pkg/controller/cheinstallation/cheinstallation_controller.go

```diff
@@ -109,14 +117,37 @@ func (r *ReconcileCheInstallation) Reconcile(request reconcile.Request) (reconci
 	reqLogger.Info("Reconciling CheInstallation")
 
 	cheInstallation := &v1alpha1.CheInstallation{}
-	err := r.client.Get(context.TODO(), request.NamespacedName, cheInstallation)
+	err := r.client.Get(context.TODO(), types.NamespacedName{
+		Name: InstallationName,
+	}, cheInstallation)
 	if err != nil {
 		if errors.IsNotFound(err) {
 			reqLogger.Info("CheInstallation not found")
 			return reconcile.Result{}, nil
 		}
 		return reconcile.Result{}, err
 	}
+	// ensure there's a finalizer, unless it's being deleted
+	if !util.IsBeingDeleted(cheInstallation) {
+		// Add the finalizer if it is not present
+		if err := r.addFinalizer(reqLogger, cheInstallation); err != nil {
+			return reconcile.Result{}, err
+		}
+	} else if util.IsBeingDeleted(cheInstallation) && util.HasFinalizer(cheInstallation, toolchainv1alpha1.FinalizerName) { // Che Installation is being deleted, but before that we should delete the Che Operator namespace explicitely
+		reqLogger.Info("Terminating CheInstallation")
+		if deleted, err := r.ensureCheClusterDeletion(reqLogger, cheInstallation); err != nil {
+			return reconcile.Result{}, r.wrapErrorWithStatusUpdate(reqLogger, cheInstallation, r.setStatusCheInstallationFailed, err, "failed to delete CheCluster resource in namespace %s", cheInstallation.Spec.CheOperatorSpec.Namespace)
+		} else if deleted {
+			return reconcile.Result{}, r.setStatusCheInstallationTerminating(cheInstallation, "deleting CheCluster resource")
+		} else {
+			// CheCluster resource as already deleted, we can now remove the finalizer on the CheInstallation
+			util.RemoveFinalizer(cheInstallation, toolchainv1alpha1.FinalizerName)
+			if err := r.client.Update(context.Background(), cheInstallation); err != nil {
+				return reconcile.Result{}, r.wrapErrorWithStatusUpdate(reqLogger, cheInstallation, r.setStatusCheInstallationTerminating, err, "failed to remove finalizer")
+			}
+			return reconcile.Result{}, nil
+		}
+	}
```

Comment:
shouldn't we return in case the `CheInstallation` is being deleted but the finalizer is not there? Otherwise, the code will keep going and start checking the resources for their existance again...

Reply:
right, let me change that

Reply:
done in f40f0f9